### DISCUSSION TITLE
5. 데이터베이스 - 트랜잭션, 동시성 제어, 회복

### DISCUSSION THREAD

**Discussion by zini9188** [Shortudy-10th/tech-for-developer#154] [질의 응답]
- 트랜잭션
- 동시성 제어
- 트랜잭션 고립 수준
- 회복
- 그 외 등등

**Comment by Jihn0118** [comment]
## Q. 데이터베이스 트랜잭션이 무엇입니까?
트랜잭션은 데이터베이스의 상태를 변화 시키기 위해 수행하는 작업 단위입니다.
(상태를 변화? -> SQL 질의어를 통해 데이터베이스에 접근하는 것)

**Reply by Jihn0118** [reply]
## Q. 데이터베이스 트랜잭션의 특징에 대해서 아는 대로 설명해주세요.

### 트랜잭션의 특징은 **원자성, 일관성, 독립성, 지속성**이 있습니다.


원자성은 
트랜잭션이 **DB에 모두 반영되거나, 혹은 모두 반영되지 않아야 되고**

일관성은 
트랜잭션의 **작업 처리 결과는 항상 일관성 있어야 하고**

독립성은 
**둘 이상의 트랜잭션이 동시에 실행되고 있을 때, 어떤 트랜잭션도 다른 트랜잭션 연산에 끼어들 수 없고,**

지속성은 
**트랜잭션이 성공적으로 완료되었으면, 결과는 영구적으로 반영**되어야 합니다.

**Comment by watty-exclamationMark** [comment]
## Q. DB 락에 대해 설명해 주세요.
트랜잭션 처리의 순서를 보장하기 위한 방법입니다.
공유락과 배타락이 있습니다.

**Comment by zini9188** [comment]
## Q. 데이터베이스의 회복 기법에 대해 설명해주세요.
데이터베이스의 회복 기법은 시스템 장애나 오류로부터 데이터베이스를 복구하는 기법으로 로그 기반 복구와 체크포인트 기반 복구가 있습니다.

**Reply by zini9188** [reply]
## Q. 로그 기반 복구와 체크포인트 기반 복구에 대해 설명해주세요.
두 기법 모두 UNDO와 REDO를 통해 데이터 일관성을 유지합니다.

로그 기반 복구는 트랜잭션 수행 중 변경된 내용을 로그에 저장하여 장애가 발생하는 경우 로그를 분석하여 트랜잭션을 재수행하거나 롤백하는 기법입니다. 

체크포인트 기법은 장애 발생 시 검사점 이전에 처리된 트랜잭션은 회복에서 제외하고
검사점 이후에 처리된 트랜잭션은 회복 작업 수행하며 검사점 이후, 장애 발생 이전에 commit이 완료된 경우 Undo를 수행하고 장애 발생 시점까지 commit되지 못한 경우 Redo를 수행합니다
